lib/tapioca/runtime/reflection.rb

```diff
@@ -168,15 +168,14 @@ def resolve_loc(locations)
         resolved_loc.absolute_path || ""
       end
 
-      sig { params(constant: Module).returns(T.nilable(String)) }
-      def constant_name_from_singleton_class(constant)
-        constant.to_s.match("#<Class:(.+)>")&.captures&.first
-      end
+      sig { params(singleton_class: Module).returns(T.nilable(Module)) }
+      def attached_class_of(singleton_class)
+        # https://stackoverflow.com/a/36622320/98634
+        result = ObjectSpace.each_object(singleton_class).find do |klass|
+          singleton_class_of(T.cast(klass, Module)) == singleton_class
+        end
```

Comment:
Also, could we add a fast path?

If `inspect` is not overridden, then we can use the old technique

Reply:
I don't think that pseudocode would work since the problem is that the `inspect` method is overriden not on the singleton class but on the original class, i.e. `#<Class:Foo>` has the default `inspect`, but `Foo` has an overriden `inspect` method. Since we don't have a handle on `Foo` to begin with, we can't check if its `inspect` method is overriden either.

Reply:
Re: comment 1: hahaha good point.

Re: any mistake there might be costly. I'd rather we keep the less performant version for now and optimize later if we see it being a problem.

Agree.